apple/internal/partials/framework_import.bzl

```diff
@@ -193,15 +203,112 @@ def _framework_import_partial_impl(
+        ctx,
+        "apple.package_symbols",
+        False,
+    )
+    if package_symbols and symbols_requested:
+        transitive_dsyms = [
+            x[AppleFrameworkImportInfo].dsym_imports
+            for x in targets
+            if AppleFrameworkImportInfo in x
+        ]
+        symbols = _generate_symbols(
+            ctx,
+            build_archs_found,
+            files_by_framework,
+            framework_binaries_by_framework,
+            transitive_dsyms,
+        )
+        bundle_files = [(
+            processor.location.archive,
+            "Symbols",
+            depset(symbols),
+        )]
+    else:
+        bundle_files = []
+
     return struct(
+        bundle_files = bundle_files,
         bundle_zips = bundle_zips,
         signed_frameworks = depset(signed_frameworks_list),
     )
 
+def _generate_symbols(
+    ctx,
+    build_archs_found,
+    files_by_framework,
+    framework_binaries_by_framework,
+    transitive_dsyms):
+    # Collect dSYM binaries and framework binaries of frameworks that don't
+    # have dSYMs
+    all_binaries = []
+    # Keep track of frameworks that provide dSYM, so that we can avoid
+    # unnecessarily extracting symbols from said frameworks' binaries
+    has_dsym_framework_basenames = sets.make()
+
+    for file in depset(transitive = transitive_dsyms).to_list():
+        # Any files that aren't Info.plist are DWARF binaries. There may be
+        # more than one binary per framework depending on how the dSYM bundle
+        # is packaged.
+        if file.basename.lower() != "info.plist":
+            all_binaries.append(file)
+            # Update the set of frameworks that provide dSYMs
+            framework_dsym_path = bundle_paths.farthest_parent(
+                file.short_path,
+                "framework.dSYM",
+            )
+            framework_dsym_basename = paths.basename(framework_dsym_path)
+            framework_basename = framework_dsym_basename.rstrip(".dSYM")
+            sets.insert(has_dsym_framework_basenames, framework_basename)
+
+    # Find binaries of frameworks that don't provide dSYMs
+    for framework_basename in files_by_framework.keys():
+        for framework_binary in framework_binaries_by_framework[framework_basename]:
+            if not sets.contains(has_dsym_framework_basenames, framework_basename):
+                all_binaries.append(framework_binary)
+
+    temp_path = paths.join("_imported_frameworks", "symbols_files")
+    symbols_dir = intermediates.directory(
+        ctx.actions,
+        ctx.label.name,
+        temp_path,
+    )
+    outputs = [symbols_dir]
+
+    commands = ["mkdir -p \"${OUTPUT_DIR}\""]
+
+    for binary in all_binaries:
+        # If dSYMs are bundled with multiple non-fat binaries, the 'symbols'
+        # command may try to extract symbols from a binary that doesn't have a
+        # slice for an architecture, but it's fine since it won't return a
+        # non-zero code in that case.
```

Comment:
Not sure if we should optimize for this case since it's very uncommon. If we do, then it might be better to extract these commands into a tool, than adding more ad-hoc bash here.

Reply:
I'm fine with not optimizing for now, that is something we can improve later. Thoughts @keith?